crates/matrix-sdk-crypto/src/store/memorystore.rs

```diff
@@ -272,47 +340,90 @@ impl CryptoStore for MemoryStore {
 
     async fn inbound_group_session_counts(
         &self,
-        _backup_version: Option<&str>,
+        backup_version: Option<&str>,
     ) -> Result<RoomKeyCounts> {
         let backed_up =
-            self.get_inbound_group_sessions().await?.into_iter().filter(|s| s.backed_up()).count();
+            if let Some(backup_order) = backup_version.and_then(|v| self.get_backup_order(v)) {
+                self.get_inbound_group_sessions_and_backup_order()
+                    .await?
+                    .into_iter()
+                    .filter(|(_, o)| *o == backup_order)
+                    .count()
+            } else {
+                0
+            };
 
         Ok(RoomKeyCounts { total: self.inbound_group_sessions.count(), backed_up })
     }
 
     async fn inbound_group_sessions_for_backup(
         &self,
-        _backup_version: &str,
+        backup_version: &str,
         limit: usize,
     ) -> Result<Vec<InboundGroupSession>> {
-        Ok(self
-            .get_inbound_group_sessions()
-            .await?
-            .into_iter()
-            .filter(|s| !s.backed_up())
-            .take(limit)
-            .collect())
+        // Given an iterator, limit it, collect it into a `Vec` and wrap it in `Ok`.
+        // The two branches of the `if` below both result in iterators, but because they
+        // are different types it is difficult to collect them into an
+        // intermediate variable, so we share code using this function instead.
+        fn take_and_collect<T: Iterator<Item = InboundGroupSession>>(
+            limit: usize,
+            it: T,
+        ) -> Result<Vec<InboundGroupSession>> {
+            Ok(it.take(limit).collect())
+        }
+
+        if let Some(requested_order) = self.get_backup_order(backup_version) {
+            // We were asked for a known backup version - only return sessions that are not
+            // already backed up in that session or a later one
+            let it =
+                self.get_inbound_group_sessions_and_backup_order().await?.into_iter().filter_map(
+                    |(session, order)| if order >= requested_order { None } else { Some(session) },
```

Comment:
as discussed, this should probably be  `==` rather than `>=`

Reply:
... except it can't in the indexeddb impl, because you can't efficiently query for rows that *don't* match a given condition.